rust/item/src/lib.rs

```diff
@@ -1679,6 +1680,146 @@ pub mod item {
         Ok(())
     }
 
+    pub fn item_class_join_namespace<'a, 'b, 'c, 'info>(
+        ctx: Context<'a, 'b, 'c, 'info, ItemClassJoinNamespace<'info>>,
+    ) -> Result<()> {
+        if !is_namespace_program_caller(&ctx.accounts.instructions.to_account_info()) {
+            return Err(error!(ErrorCode::UnauthorizedCaller));
+        }
+
+        let item_class = &mut ctx.accounts.item_class;
+
+        let namespaces = match item_class.namespaces.clone() {
+            Some(namespaces) => namespaces,
+            None => return Err(error!(ErrorCode::FailedToJoinNamespace)),
+        };
+
+        let mut joined = false;
+        let mut new_namespaces = vec![];
+        for mut ns in namespaces {
+            if ns.namespace == anchor_lang::solana_program::system_program::id() && !joined {
+                ns.namespace = ctx.accounts.namespace.key();
+                ns.index = None;
+                ns.inherited = InheritanceState::NotInherited;
+                joined = true;
+                new_namespaces.push(ns);
+            } else {
+                new_namespaces.push(ns);
+            }
+        }
+        if !joined {
+            return Err(error!(ErrorCode::FailedToJoinNamespace));
+        }
+        item_class.namespaces = Some(new_namespaces);
+
+        Ok(())
+    }
+
+    pub fn item_class_leave_namespace<'a, 'b, 'c, 'info>(
+        ctx: Context<'a, 'b, 'c, 'info, ItemClassLeaveNamespace<'info>>,
+    ) -> Result<()> {
+        if !is_namespace_program_caller(&ctx.accounts.instructions.to_account_info()) {
+            return Err(error!(ErrorCode::UnauthorizedCaller));
+        }
+
+        let item_class = &mut ctx.accounts.item_class;
+
+        let namespaces = match item_class.namespaces.clone() {
+            Some(namespaces) => namespaces,
+            None => return Err(error!(ErrorCode::FailedToLeaveNamespace)),
+        };
+
+        let mut left = false;
+        let mut new_namespaces = vec![];
+        for mut ns in namespaces {
+            if ns.namespace == ctx.accounts.namespace.key() && !left {
+                // if the artifact is still cached, error
+                if ns.index != None {
+                    return Err(error!(ErrorCode::FailedToLeaveNamespace));
+                };
+                ns.namespace = anchor_lang::solana_program::system_program::id();
+                ns.inherited = InheritanceState::NotInherited;
+                left = true;
+                new_namespaces.push(ns);
+            } else {
+                new_namespaces.push(ns);
+            }
```

Comment:
Educational question, what is `ns.index` when leaving namespace?

Reply:
When an artifact joins a namespace it is not cached by default, so index will remain `None`. When `cache_artifact` is called, this will change to the index number in which it is cached. There are 10 indexes per namespace each can hold 100 pubkeys, we do not allow an artifact which is still cached to leave a namespace.